gh-70795: Rework RLock documentation (#103853)
python/cpython: Doc/library/threading.rst

locks.  In the locked state, some thread owns the lock; in the unlocked state,
no thread owns it.

> To lock the lock, a thread calls its :meth:`~RLock.acquire` method.  To unlock

Reviewer: ```suggestion
To lock the lock, a thread calls the lock's :meth:`~RLock.acquire` method.  To unlock
```

Avoid any ambiguity about whether "it" is the thread or the lock.

> the lock, a thread calls its :meth:`~Lock.release` method. Reentrant locks

Reviewer: ```suggestion
Threads call a lock's :meth:`~RLock.acquire` method to lock it,
and its :meth:`~Lock.release` method to unlock it. Reentrant locks
```

Alternatively, this would fix the above issue and also reduce duplication and redundancy.

> support the :ref:`context management protocol <with-locks>` so ```with``` can be used.

Reviewer: ```suggestion
support the :ref:`context management protocol <with-locks>` so :keyword:`with` can be used.
```

Fix syntax mistake (triple instead of double backticks); format and link `with`.

Reviewer: Also, `with` can be used how? I'm guessing maybe

```suggestion
support the :ref:`context management protocol <with-locks>` so :keyword:`with` can be used
to acquire and release the lock automatically in a block of code.
```

RLock's :meth:`~RLock.acquire`/:meth:`~RLock.release` call pairs may be nested,
unlike Lock's :meth:`~Lock.acquire`/:meth:`~Lock.release`. Only the final
:meth:`~RLock.release` (the :meth:`~Lock.release` of the outermost pair) resets
the lock to an unlocked state and allows another thread blocked in
:meth:`~RLock.acquire` to proceed.

> :meth:`~RLock.acquire`/:meth:`~RLock.release` must be in pairs: each acquire

Reviewer: ```suggestion
:meth:`~RLock.acquire`/:meth:`~RLock.release` must be used in pairs: each acquire
```

Minor clarification

must have a release in the thread that has acquired the lock. Failing to

> release as many times the lock has required will lead to deadlock.

Reviewer: ```suggestion
call release as many times the lock has been acquired can lead to deadlock.
```

Fix typo and apply Greg's suggestions

> release as many times the lock has required will lead to deadlock.

Reviewer 2: wording, perhaps:

... call release as many times as the lock has been acquired can lead to deadlock.

Reviewer: Yup; I added this in my suggestion

locks.  In the locked state, some thread owns the lock; in the unlocked state,
no thread owns it.

Threads call a lock's :meth:`~RLock.acquire` method to lock it,
and its :meth:`~Lock.release` method to unlock it. Reentrant locks
support the :ref:`context management protocol <with-locks>` so :keyword:`with` can be used

> to acquire and release the lock automatically in a block of code.

Reviewer: ```suggestion
and its :meth:`~Lock.release` method to unlock it.

.. note::

  Reentrant locks support the :ref:`context management protocol <with-locks>`,
  so it is recommended to use :keyword:`with` instead of manually calling
  :meth:`~RLock.acquire` and :meth:`~RLock.release`
  to handle acquiring and releasing the lock for a block of code.
```

Call this out in a note

.. seealso::

>    :ref:`with-locks`.

Reviewer: ```suggestion
   :ref:`Using RLock as a context manager <with-locks>` is recommended
   over manual :meth:`!acquire` and :meth:`release` calls whenever practical.
```

Here's how I would phrase this, and also move it down past the first summary sentence in `RLock.acquire()` ("Acquire a lock, blocking or non-blocking")

      Acquire a lock, blocking or non-blocking.

      When invoked with the *blocking* argument set to ``True`` (the default):

         1. If another thread owns the lock, block until we are able to acquire

>             lock, or *timeout*, if set.

Reviewer: ```suggestion
            lock, or *timeout*, if set to a positive float value.
```

Be more explicit here and incorporate the additional information below.

         2. If no thread owns the lock, acquire the lock and return immediately.

         3. If the same thread owns the lock, acquire the lock again, and
            return immediately. This is the difference between ``Lock`` and

>             ``RLock``. Internally, the "recursion level" is atomically

Reviewer 2: Lets not mention the internal implementation details.  get rid of the "Internally, ..." sentence.

Reviewer: Yeah, noticed that as well. Implemented in my suggestions

Author: Removed.

Note: we talk about other internal details in the beginning of the RLock, as well as the doc for `release`, that this PR otherwise did not touch.

Reviewer: Seems like this is resolved, so I'll mark this as resolved and merge this, and we can iterate further if necessary in a followup. Thanks!

>             incremented to allow this.

Reviewer: This should be a bulleted rather than ordered list, as the latter implies a sequence of steps with an explict order rather than multiple different options/paths with no fixed, predetermined order.

Additionally, I would suggest swapping the first and second item, since the current case "2" is the simplest.

Author: Fixed.

>             return immediately. This is the difference between ``Lock`` and
>             ``RLock``. Internally, the "recursion level" is atomically
>             incremented to allow this.

Reviewer: ```suggestion
            return immediately. This is the difference between :class:`Lock` and
            :class:`!RLock`; :class:`Lock` handles this case the same as the previous,
            blocking until the lock can be acquired.
```

Be much more explicit and clear as to how these two differ. Also, remove the last sentence, which is an internal implementation detail.

>       When invoked with the *blocking* argument set to ``False``, do not block.
>       Without blocking:

Reviewer: ```suggestion
      When invoked with the *blocking* argument set to ``False``:
```

Just jump directly to what it actually means, like the previous ``True`` case

         1. If another thread owns the lock, return immediately.

         2. If no thread owns the lock, acquire the lock and return immediately.

>          3. If the same thread owns the lock. acquire the lock again and return

Reviewer: ```suggestion
         3. If the same thread owns the lock, acquire the lock again and return
```

Fix typo

>          3. If the same thread owns the lock. acquire the lock again and return

Reviewer 2: comma instead of .

Reviewer: Yup, fixed in my suggestion

>             immediately.

Reviewer: Same here—this should be a bulleted list, with the simplest case (2.) first.

>       When invoked with the floating-point *timeout* argument set to a positive

Reviewer: ```suggestion
```

Remove this redundant sentence; incorporated above.

>       value, block for at most the number of seconds specified by *timeout*.
> 
>       In all cases, if the thread was able to acquire the lock, return ``True``.

Reviewer: ```suggestion
      In all cases, if the thread was able to acquire the lock, return ``True``.
```

As above

      If the thread was unable to acquire the lock (i.e. if *blocking* is False, or

>       is reached timeout) return ``False``.

Reviewer 2: ... (i.e. if not blocking or the timeout was reached) ...

>       is reached timeout) return ``False``.

Reviewer: ```suggestion
      *timeout* is reached or is non-positive) return ``False``.
```

Fix grammar, add formatting and clarify case of timeout being non-positive

      If called multiple times, failing to call :meth:`~RLock.release` as many times

>       may lead to deadlock. Consider using ``RLock`` as a context manager rather than

Reviewer: ```suggestion
      may lead to deadlock. Consider using :class:`!RLock` as a context manager rather than
```

>       calling acquire/release directly.

Reviewer: I suggest making this a `.. seealso::` block after the summary line rather than buried at the end.

Author: Fixed.

      Acquire a lock, blocking or non-blocking.

      .. seealso::

>          :ref:`Using RLock as a context manager <with-locks>` is recommended
>             over manual :meth:`!acquire` and :meth:`release` calls whenever practical.

Reviewer: ```suggestion
         :ref:`Using RLock as a context manager <with-locks>`
            Recommended over manual :meth:`!acquire` and :meth:`release` calls
            whenever practical.
```

      Only call this method when the calling thread owns the lock. A
      :exc:`RuntimeError` is raised if this method is called when the lock is

>       un-acquired.

Reviewer: ```suggestion
      not acquired.
```

Un-acquired would be interpreted as verb here, where this is in fact a noun (a state)

>       un-acquired.

Reviewer 2: "if this method is called on a non-acquired lock." perhaps?

Reviewer: I changed it to just "if this method is called when the lock is not acquired" but we can revise it further if needed.
